### PR TITLE
android x support

### PR DESCRIPTION
**Pull request by shiraji**
See #488 

~We need to wait bintray support gradle 4.6 or release this alpha version manually ✋~

### TODO

- [x] Support androidx
- [x] Drop support v4
- [x] Handle the problem with bintray-release

### NOT TODO for this PR

* Drop native fragment support

### CODE REVIEW

**Comment by shiraji** [comment]
It seems dropping native fragment support makes diff huge. Do you guys want me to split PR? @hotchemi @mannodermaus

**Comment by hotchemi** [comment]
yea it's better I guess!

**Comment by shiraji** [comment]
I will create new PR for these commits.

f745322
cb7d58b
edd66c4

**Comment by shiraji** [comment]
Ah, one more thing, do you want me to send PR for non-master branch like `version-4` or somthing?

**Comment by hotchemi** [comment]
To master is ok! But to merge this PR we need to deal with bintray-release plugin.. 🤔

File: .travis.yml [hunk 1]
```diff
@@ -20,7 +20,9 @@ jdk:
   - oraclejdk8
 
 before_install:
-- echo "sdk.dir=$ANDROID_HOME" > local.properties
+  - echo "sdk.dir=$ANDROID_HOME" > local.properties
+  - mkdir "$ANDROID_HOME/licenses" || true
+  - cp ./android-sdk-licenses/* "$ANDROID_HOME/licenses/"
```
Comment (shiraji) [inline .travis.yml]: We might not need this change because of bintray-release plugin

File: gradle.properties [hunk 2]
```diff
@@ -29,11 +30,13 @@ ROBOLECTRIC_VERSION     = 3.3.2
...
+COMPILE_SDK_VERSION     = android-28
+BUILD_TOOLS_VERSION     = 28.0.0-rc2
+TARGET_SDK_VERSION      = 28
 MIN_SDK_VERSION         = 14
 
 # Gradle parameters
 org.gradle.daemon       = true
 org.gradle.jvmargs      = -XX:MaxPermSize=1024m -XX:+CMSClassUnloadingEnabled -XX:+HeapDumpOnOutOfMemoryError -Xmx2048m
+android.useAndroidX=true
+android.enableJetifier=true
```
Comment (shiraji) [inline gradle.properties]: Formatting

File: lint/src/test/java/permissions/dispatcher/CallOnRequestPermissionsResultDetectorTest.kt [hunk 3]
```diff
@@ -9,6 +9,7 @@ import permissions.dispatcher.Utils.onNeedsPermission
 import permissions.dispatcher.Utils.onRationaleAnnotation
 import permissions.dispatcher.Utils.runtimePermission
 
+
```
Comment (shiraji) [inline lint/src/test/java/permissions/dispatcher/CallOnRequestPermissionsResultDetectorTest.kt]: What's this?

**Comment by shiraji** [comment]
> we need to deal with bintray-release plugin

3 options I came up:
* Wait until bintray-release plugin support gradle 4.6
* Manual upload 😉 
* Migrate to [gradle-bintray-plugin](https://github.com/bintray/gradle-bintray-plugin)

If bintray-release support gradle 4.6 is best however because they don't even support gradle 4.5, I would like to drop this plugin and migrate to official one.

**Comment by hotchemi** [comment]
@shiraji Thx! seemingly https://github.com/bintray/gradle-bintray-plugin/pull/230 has not been merged yet so I'm not sure official plugin works well with Gradle4.6. But I'm definitely for adopting official one!

**Comment by shiraji** [comment]
@hotchemi According to https://github.com/bintray/gradle-bintray-plugin/issues/239, people already use Gradle 4.6 or higher. Let me try that we can at least compile it.

**Comment by hotchemi** [comment]
Oh noted. Thx!

**Comment by dharrigan** [comment]
Hi,

Any progress on getting a new release out that supports Android X? I have a new project that fails to compile because v4 compat (ActivityCompat) is no longer available on the generated Dispatcher.

Thank you.

**Comment by shiraji** [comment]
I was working on replacing bintray-release with gradle-bintray-plugin. I'm not sure how I can test this fix at this moment.

**Comment by hotchemi** [comment]
Ah sorry I jus started working on from today @shiraji may be I'm a person to test🙇

**Comment by shiraji** [comment]
I'll push my code asap which is what I think it should. Then, you can test it. I really don't know if it works or not, thou

**Comment by dharrigan** [comment]
Happy to test out the new artifact when deployed :-)

-=david=-

File: .travis.yml (hunk 1 above)
Comment (hotchemi) [reply .travis.yml]: 👀

**Comment by hotchemi** [comment]
Thank you for your great work @shiraji !
I'll release 4.0.0-alpha1 tonight with some tests just in case!

**Comment by shiraji** [comment]
@dharrigan Now, 4.0.0-SNAPSHOTS is ready. Please give us your feedback!

**Comment by codelishaohui** [comment]
where is the 4.0.0-alpha1 ?  ：）

**Comment by hotchemi** [comment]
Tonight!

**Comment by dharrigan** [comment]
Hi,

Well, testing...and it doesn't appear to work! 

```
Android Studio 3.2 Beta 2
Build #AI-181.5281.24.32.4860949, built on June 25, 2018
JRE: 1.8.0_152-release-1136-b04 amd64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
Linux 4.15.0-20-generic
```
```
implementation("com.github.hotchemi:permissionsdispatcher:4.0.0-SNAPSHOT") {
    exclude module: "support-v13"
}
```
```
annotationProcessor "com.github.hotchemi:permissionsdispatcher-processor:4.0.0-SNAPSHOT"
```

```
@RuntimePermissions
class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    @NeedsPermission(ACCESS_FINE_LOCATION)
    fun scanForDevices() {

    }

    @OnShowRationale(ACCESS_FINE_LOCATION)
    fun showRationaleForAccessFineLocation(request: PermissionRequest) {
        showRationaleDialog(R.string.permission_access_fine_location_rationale, request)
    }

    @OnPermissionDenied(ACCESS_FINE_LOCATION)
    fun onAccessFineLocationDenied() {
        makeText(this, R.string.permission_access_fine_location_denied, LENGTH_SHORT).show()
    }

    @OnNeverAskAgain(ACCESS_FINE_LOCATION)
    fun onAccessFineLocationNeverAskAgain() {
        makeText(this, R.string.permission_access_fine_location_never_ask_again, LENGTH_SHORT).show()
    }

    private fun showRationaleDialog(@StringRes messageResId: Int, request: PermissionRequest) {
        AlertDialog.Builder(this)
                .setPositiveButton(R.string.button_allow) { _, _ -> request.proceed() }
                .setNegativeButton(R.string.button_deny) { _, _ -> request.cancel() }
                .setCancelable(false)
                .setMessage(messageResId)
                .show()
    }

}
```

When doing a full build, no generated classes seem to be created, i.e., no `scanForDevicesWithPermissionCheck`. The `generatedJava` directory is empty except for the dispatcher, R class.

-=david=-

**Comment by shiraji** [comment]
@dharrigan Sorry if I confused you but it's not released, yet.

**Comment by hotchemi** [comment]
@dharrigan thx for checking. are there any error log? I'm gonna test!

**Comment by dharrigan** [comment]
Hi,

No error logs - compilation succeeds normally.

**Comment by dharrigan** [comment]
Hi,

Actually, discovered the error.

I had `annotationProcessor` rather than `kapt` in my build.gradle - d'oh!

However, even with the snapshot 4.0 library (and the generated classes being generated), It's still complaining about `Unresolved reference: ActivityCompat` in MainActivityPermissionsDispatcher.

**Comment by hotchemi** [comment]
I'm testing but seems working? 👁

**Comment by dharrigan** [comment]
Hi,

Perhaps I'm missing something? I did a clean clone of the project, pulled down all the dependencies again, ensured I was using the 4.0.0-SNAPSHOT, did a build from the command line and I get the same error as in 3.3.1, i.e.,

`Unresolved reference: ActivityCompat` in `MainActivityPermissionsDispatcher.kt`

**Comment by hotchemi** [comment]
@dharrigan could you try 4.0.0-alpha? 👀 Seems we failed to release library module.

**Comment by dharrigan** [comment]
Hi,

Seems like the dependency fails:

Inside 

`https://bintray.com/hotchemi/maven/download_file?file_path=com%2Fgithub%2Fhotchemi%2Fpermissionsdispatcher%2F4.0.0-alpha1%2Fpermissionsdispatcher-4.0.0-alpha1.pom`

```
<dependency>
<groupId>PermissionsDispatcher</groupId>
<artifactId>annotation</artifactId>
<version>unspecified</version>
<scope>compile</scope>
</dependency>
```
When added as a depdency in build.gradle:

```
Failed to resolve: PermissionsDispatcher:annotation:unspecified
Open File
Show in Project Structure dialog
```

**Comment by hotchemi** [comment]
OMG sorry about that.

**Comment by hotchemi** [comment]
Umm why it happens..

**Comment by hotchemi** [comment]
@permissions-dispatcher/maintainers does anyone have an idea why this happens.. 🤔 
https://github.com/permissions-dispatcher/PermissionsDispatcher/pull/494#issuecomment-404079580

**Comment by shiraji** [comment]
How we can generate pom file for `library` module?

**Comment by hotchemi** [comment]
@shiraji https://github.com/permissions-dispatcher/PermissionsDispatcher/blob/92540b7b8ca92ca266d7e6cc7dff44e2e5ec7c18/scripts/release.sh#L5

I'll send needed info privately!

**Comment by hotchemi** [comment]
@dharrigan we've fixed by https://github.com/permissions-dispatcher/PermissionsDispatcher/pull/500, can you try again?

**Pull request merged**